lib/annotations2triannon/open_annotation.rb

```diff
@@ -261,6 +254,33 @@ def to_ttl
       @graph.dump(:ttl, standard_prefixes: true)
     end
 
+    private
+
+    # Sets the annotation ID as an RDF::URI from id, but if the id is nil, it
+    # will try to get the ID from the graph and, as a last resort, it will
+    # generate a unique ID using a UUID.
+    # @param id [UUID|URI|String] to identify an open annotation
+    def set_id(id)
+      raise ArgumentError, 'id cannot be an empty String' if (id.instance_of?(String) && id.empty?)
+      if id.nil?
+        # Try to get the ID from the graph or use a UUID
+        q = [nil, RDF.type, OA.Annotation]
+        @id = @graph.query(q).collect {|s| s.subject }.first || RDF::URI.parse(UUID.generate)
```

Comment:
URIs will be overwritten with the dereferenceable URI from Fedora4. So blank node or resource with a URI makes no difference *as input* compared to what is persisted and made available.  Triannon at some point should maintain those URIs (see provenance ticket) when they're supplied, but it's not a requirement for beta.